### PR TITLE
Parse to extract all scopes and bindings.

### PR DESCRIPTION
**Pull request by yurydelendik**
Associated Issue: #3817 

### Summary of Changes

* allows to send source text independent from parsing of the scopes
* parsing scopes to extract positions and its scope of every symbol in the source code

### CODE REVIEW

File: src/utils/parser/getScopes.js [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+        parent = savedParent;
+        savedParents.delete(path);
+      }
+    }
+  });
+  return _toGeneratedScopes(parent.scopes, source.id);
+}
+
+function _compareLocations(a: Location, b: Location): number {
+  // $FlowIgnore
```
Comment (codehag) [inline src/utils/parser/getScopes.js]: why are we ignoring here?

File: src/utils/parser/getScopes.js [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+  const { sourceId } = location;
+  let parsedScopes = parsedScopesCache.get(sourceId);
+  if (!parsedScopes) {
+    parsedScopes = _parseJSScopes(getSource(sourceId));
+    parsedScopesCache.set(sourceId, parsedScopes);
+  }
+
+  // Find inner most
+  let search = parsedScopes;
+  let found = [];
```
Comment (codehag) [inline src/utils/parser/getScopes.js]: this should be const

File: src/utils/parser/getScopes.js [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        let scope = parent;
+        if (hasLet) {
+          savedParents.set(path, parent);
+          scope = _createTempScope("Block", parent, location);
+          parent = scope;
+        }
+        return;
+      }
+      if (path.isVariableDeclaration()) {
+        let hoistAt = parent;
```
Comment (codehag) [inline src/utils/parser/getScopes.js]: this doesnt need to be assigned to parent, would make sense to assign this to null, or turn the if statement into a function that returns the hoistAt value or null

File: src/utils/parser/getScopes.js [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+      }
+      if (path.isForInStatement() || path.isForStatement()) {
+        const init = tree.init || tree.left;
+        let scope = parent;
+        if (
+          isNode(init, "VariableDeclaration") &&
+          (init.kind === "let" || init.kind === "const")
+        ) {
+          savedParents.set(path, parent);
+          scope = _createTempScope("For", parent, location);
```
Comment (codehag) [inline src/utils/parser/getScopes.js]: scope can be declared here (as it is in other places)
`const scope = _createTempScope ...`

File: src/utils/parser/getScopes.js [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+        }
+        tree.params.forEach(param => {
+          _parseDeclarator(param, scope, "param");
+        });
+        parent = scope;
+        return;
+      }
+      if (path.isForInStatement() || path.isForStatement()) {
+        const init = tree.init || tree.left;
+        let scope = parent;
```
Comment (codehag) [inline src/utils/parser/getScopes.js]: we don't need this assignment

File: src/utils/parser/getScopes.js [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+    return undefined;
+  }
+  return children.map(scope => {
+    let bindings = Object.create(null);
+    Object.keys(scope.names).forEach(n => {
+      switch (scope.names[n].kind) {
+        case "var":
+        case "let":
+        case "const":
+        case "param":
```
Comment (codehag) [inline src/utils/parser/getScopes.js]: TIL that this is faster than using regex!

File: src/utils/parser/getScopes.js [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+function _toGeneratedScopes(
+  children: TempScope[],
+  sourceId: SourceId
+): ?(GeneratedScope[]) {
+  if (!children || children.length === 0) {
+    return undefined;
+  }
+  return children.map(scope => {
+    let bindings = Object.create(null);
+    Object.keys(scope.names).forEach(n => {
```
Comment (codehag) [inline src/utils/parser/getScopes.js]: this should probably be a reduce. I would expect forEach if we were triggering side effects.

```javascript
const bindings = Object.keys(scope.names).reduce((bindings, n) => {
  const name = scope.names[n]; // or something
  switch (name.kind) {
    case "var":
    case "let":
    case "const":
    case "param":
      const binding = name.refs.map(
         location => _fromBabelLocation(location, sourceId) 
      );
      return { ...bindings, [n]: binding };
    }
    return bindings;
}
```

this way we can also factor this out into its own function

File: src/utils/parser/getScopes.js [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+  }
+}
+
+function _recordIdentifierPosition(
+  scope: TempScope,
+  name: string,
+  loc: BabelLocation
+) {
+  for (let s = scope; s; s = s.parent) {
+    if (name in s.names) {
```
Comment (codehag) [inline src/utils/parser/getScopes.js]: what is happening here? is it looking over s.names?

File: src/utils/parser/getScopes.js [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+        return;
+      }
+      if (path.isBlockStatement()) {
+        const hasLet = tree.body.some(node => {
+          if (!isNode(node, "VariableDeclaration")) {
+            return false;
+          }
+          return node.kind === "let" || node.kind === "const";
+        });
+        let scope = parent;
```
Comment (codehag) [inline src/utils/parser/getScopes.js]: scope doesnt need to be declared here.

File: src/utils/parser/getScopes.js [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+    }
+  }
+}
+
+function _toGeneratedScopes(
+  children: TempScope[],
+  sourceId: SourceId
+): ?(GeneratedScope[]) {
+  if (!children || children.length === 0) {
+    return undefined;
```
Comment (codehag) [inline src/utils/parser/getScopes.js]: perhaps we can return an empty array, so that we are not changing the type of the return value

File: src/utils/parser/getScopes.js [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+        const hasLet = tree.body.some(node => {
+          if (!isNode(node, "VariableDeclaration")) {
+            return false;
+          }
+          return node.kind === "let" || node.kind === "const";
+        });
+        let scope = parent;
+        if (hasLet) {
+          savedParents.set(path, parent);
+          scope = _createTempScope("Block", parent, location);
```
Comment (codehag) [inline src/utils/parser/getScopes.js]: can be declared with const

File: src/utils/parser/getScopes.js [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+  if (!parsedScopes) {
+    parsedScopes = _parseJSScopes(getSource(sourceId));
+    parsedScopesCache.set(sourceId, parsedScopes);
+  }
+
+  // Find inner most
+  let search = parsedScopes;
+  let found = [];
+  while (search) {
+    let foundOne = search.some(s => {
```
Comment (codehag) [inline src/utils/parser/getScopes.js]: search.some stops execution after one is found; but we are modifying the found array here and the search object. perhaps what we want here again is reduce?

```javascript
const found = search.reduce((found, s) => {
  if (
    _compareLocations(s.start, location) <= 0 &&
    _compareLocations(location, s.end) < 0
  ) {
    return [s, ...found];
  }
  return found;
})
```

you can also factor that out into an independant function and do (imagining that the reduce is its own function, and the map is another):
```javascript
return search.reduce(findLocations).map(createScope);
```

File: src/utils/parser/sources.js [hunk 13]
```diff
@@ -0,0 +1,25 @@
+// @flow
+
+import type { Source, SourceId } from "debugger-html";
+
+var cachedSources = new Map();
```
Comment (codehag) [inline src/utils/parser/sources.js]: why do we want to use var here? would let work?

File: src/utils/parser/getScopes.js [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+  });
+  return _toGeneratedScopes(parent.scopes, source.id);
+}
+
+function _compareLocations(a: Location, b: Location): number {
+  // $FlowIgnore
+  return a.line == b.line ? a.column - b.column : a.line - b.line;
+}
+
+var parsedScopesCache = new Map();
```
Comment (codehag) [inline src/utils/parser/getScopes.js]: we are using var here, is there a particular reason? i would expect let to be used in this case

File: src/utils/parser/sources.js [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+
+export function setSource(source: Source) {
+  cachedSources.set(source.id, source);
+}
+
+export function getSource(sourceId: SourceId): Source {
+  if (!cachedSources.has(sourceId)) {
+    throw new Error(`${sourceId} was not provided.`);
+  }
+  // $FlowIgnore
```
Comment (codehag) [inline src/utils/parser/sources.js]: why are we ignoring here?

File: src/utils/parser/tests/getScopes.spec.js [hunk 16]
```diff
@@ -0,0 +1,45 @@
+import getScopes from "../getScopes";
+import { setSource } from "../sources";
+
+describe("Parser.getScopes", () => {
+  it("parses simple generated/minified es5 source", () => {
+    const source = {
```
Comment (codehag) [inline src/utils/parser/tests/getScopes.spec.js]: nit: we can use the parser's test helper and save this in its own file i think

its used via
```javascript
import { getSource } from "../helpers";

//...
const source = getSource("getScopes/source1");
```

File: src/utils/parser/getScopes.js [hunk 17]
```diff
@@ -0,0 +1,280 @@
...
+      start: _fromBabelLocation(scope.loc.start, sourceId),
+      end: _fromBabelLocation(scope.loc.end, sourceId),
+      type: scope.type.toLowerCase(),
+      bindings,
+      children: _toGeneratedScopes(scope.scopes, sourceId)
+    };
+  });
+}
+
+function _parseJSScopes(source: Source) {
```
Comment (codehag) [inline src/utils/parser/getScopes.js]: what do the `_` signify in these function names?

Comment (jasonLaster) [reply src/utils/parser/getScopes.js]: probably private

Comment (codehag) [reply src/utils/parser/getScopes.js]: thats what i thought, but wouldn't it be private by default, since it isnt exported?

File: src/utils/parser/getScopes.js (hunk 8 above)
Comment (yurydelendik) [reply src/utils/parser/getScopes.js]: it checks in `name` is a property in s.names dictionary. related to the `Object.create(null)` above

File: src/utils/parser/getScopes.js (hunk 10 above)
Comment (yurydelendik) [reply src/utils/parser/getScopes.js]: That might cause of creation many empty leaf objects.

File: src/utils/parser/sources.js (hunk 15 above)
Comment (yurydelendik) [reply src/utils/parser/sources.js]: Even if I did the check above, flow is convinced cacheSources.get can be undefined. Changing to ((x: any): Source) cast.

**Comment by yurydelendik** [comment]
Moved scope parsing related code to the https://github.com/devtools-html/devtools-core/pull/656

File: src/utils/parser/getScopes.js [hunk 18]
```diff
@@ -0,0 +1,280 @@
...
+    _parseDeclarator(declaratorId.argument, targetScope, kind);
+  }
+}
+
+function _recordIdentifierPosition(
+  scope: TempScope,
+  name: string,
+  loc: BabelLocation
+) {
+  for (let s = scope; s; s = s.parent) {
```
[the remark this answers is not in the document]
Comment (yurydelendik) [reply src/utils/parser/getScopes.js]: The recursive functions (even with tails call) are not really efficient in comparison with just a loop. I recommend to not introduce any unneeded recursion for simple algorithms, and in this case simple loop easier to read.

File: src/utils/parser/sources.js [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+export function getSource(sourceId: SourceId): Source {
+  if (!cachedSources.has(sourceId)) {
+    throw new Error(`${sourceId} was not provided.`);
+  }
+  return ((cachedSources.get(sourceId): any): Source);
+}
+
+export function clearSources() {
+  cachedSources = new Map();
+}
```
Comment (jasonLaster) [inline src/utils/parser/sources.js]: almost all of our APIs now expect to get a `source`. 

I agree that this is a performance issue, but I think we should make the fix in a separate PR, that touches all of the spots.

https://github.com/devtools-html/debugger.html/blob/master/src/utils/parser/steps.js#L6-L8
https://github.com/devtools-html/debugger.html/blob/master/src/utils/parser/getSymbols.js#L335-L336

#3953

**Comment by yurydelendik** [comment]
Depends on #4008
